File: python/crosswordMaker.py
import regex as re
from crosswordMaker import Crossword
from crosswordSolver import crosswordSolver


class Crossword(object):

    # represents all the available characters for this puzzle
    characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ 1234567890!"#%&,<=>@~_`;:'

    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.clues = dict(
            [('X' + str(i), [])
                for i in xrange(self.width)]
            + 
            [('Y' + str(j), [])
                for j in xrange(self.height)]
        )
        self.solutions = ['' for _ in xrange(self.width * self.height)]
        self.possibleMatches = [self.characters
                for _ in xrange(self.width * self.height)]
        
        
    def addClue(self, cell, exp):
        if cell[0] is not 'X' and cell[0] is not 'Y':
            raise ValueError('There are only two directions, X and Y.')
        elif int(cell[1]) < 0 or (cell[0] is 'X' and int(cell[1]) >= self.width) or (cell[0] is 'Y' and int(cell[1]) >= self.height):
            raise ValueError('Index out of bounds')
            
        self.clues[cell].append(exp);
        

File: js/crosswordMaker.js

//const escapeStringRegexp = require('escape-string-regexp');

//only characters tested by game
let characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ 1234567890!"#%&,<=>@~_`;:'


function Crossword(width, height) {
    this.width = width;
    this.height = height;

    this.clues = this.cluesMaker(height, width);
    this.possibleMatches = Array(height * width).fill(characters);
    this.solutions = Array(height * width).fill('');
}

Crossword.prototype.cluesMaker = function (height, width) {
    let clues = {};
    for (let i = 0; i < width; i++) {
        let key = 'X' + i;
        clues[key] = [];
    }

    for (let j = 0; j < height; j++) {
        let key = 'Y' + j;
        clues[key] = [];
    }

    return clues;
}


//exp must be input as String.raw`template literal`--> Is there any way to change this?
//cell is a string 
Crossword.prototype.addClue = function (cell, exp) {
    if (cell[0] !== 'X' && cell[0] !== 'Y') {
        throw Error('There are only two directions, X and Y.')
    }

    else if (cell[1] < 0 || (cell[0] === 'X' && cell[1] >= this.width) || (cell[0] === 'Y' && cell[1] >= this.height)) {
        throw Error('Index out of bounds')
    } 

    this.clues[cell].push(exp);

    }


module.exports = Crossword;